applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_shape_fraction_optimization.py

```diff
@@ -0,0 +1,620 @@
+                else:
+                    gradient_i[0] /= penalty_value**2
+                    gradient_i[1] /= penalty_value**2
+                    gradient_i[2] /= penalty_value**2
+
+            penalty_gradient[node.Id] = gradient_i
+
+        KM.Logger.PrintInfo("ShapeFractionOptimization", "Time needed for calculating the penalty gradient = ",round(timer.time() - startTime,2),"s")
+
+        return penalty_gradient
+
+    def __getQuantile(self):
+
+        nodal_variable = KM.KratosGlobals.GetVariable("SHAPE_CHANGE")
+        shape_change = ReadNodalVariableToList(self.design_surface, nodal_variable)
+
+        shape_change_norm = []
+
+        for i in range(int(len(shape_change)/3)):
+            shape_change_norm.append(cm.Norm2(shape_change[3*i:3*i+3]))
+
+        shape_change_norm.sort()
+
+        index = round(len(shape_change_norm)*(1-self.max_shape_fraction))
+
+        quantile = shape_change_norm[index]
+
+        KM.Logger.PrintInfo("ShapeFractionOptimization: Shape change quantile {}: {}".format((1-self.max_shape_fraction), quantile))
+
+        return quantile
+
+    def __computeShapeFraction(self):
+
+        n_active = 0
+        for node in self.design_surface.Nodes:
+            shape_change = node.GetSolutionStepValue(KSO.SHAPE_CHANGE)
+            norm = cm.Norm2(shape_change)
+
+            if norm > self.frac_tolerance:
+                n_active += 1
+
+        total_nodes = len(self.design_surface.Nodes)
+
+        shape_fraction = n_active / total_nodes
+
+        return shape_fraction
+
+    # --------------------------------------------------------------------------
+    def __checkInnerConvergence(self, objective_value):
+
+        tolerance = self.inner_tolerance
+
+        is_converged = False
+        if self.previous_objective_value is None:
+            self.reference_objective_value = objective_value
+            self.previous_objective_value = objective_value
+            self.relative_change = 0
+        elif self.inner_step >= self.max_inner_steps:
+            self.previous_objective_value = objective_value
+            is_converged = True
+        else:
+            self.relative_change = (objective_value - self.previous_objective_value) / self.reference_objective_value
+            KM.Logger.PrintInfo("ShapeFractionOptimization", "relative_change: {}".format(self.relative_change))
+            if abs(self.relative_change) < tolerance:
+                is_converged = True
+
+            self.previous_objective_value = objective_value
+
+        if is_converged:
+            self.inner_step = 0
+            self.reference_objective_value = objective_value
+
+        return is_converged
+
+    # --------------------------------------------------------------------------
+    def __updatePenaltyFactor(self):
+
+        if self.penalty_method == "exterior" and self.penalty_value > 0.0:
+            self.penalty_factor *= 1.25 # increase penalty factor by 25%
```

Comment:
*= 1.25 # increase penalty factor by 25%, it can be a variable with default value 1.25 . I think it is named "gamma" in Venderplaats book.